examples/nodejs/src/experimental-features/applicable-inputs.ts

```diff
@@ -0,0 +1,512 @@
+   * What is the environment to apply the inputs
+   */
+  environment: Environment;
+  /**
+   * What is the new state after applying an action and reducing until quiescent
+   */
+  reducedState: MarloweState;
+  /**
+   * What is the new contract after applying an action and reducing until quiescent
+   */
+  reducedContract: Contract;
+  /**
+   * What warnings were produced while applying an action
+   */
+  warnings: TransactionWarning[];
+  /**
+   * What payments were produced while applying an action
+   */
+  payments: Payment[];
+}
+
+interface CanNotify {
+  type: "Notify";
+  policyId: PolicyId;
+  applyAction(): AppliedActionResult;
+}
+
+interface CanDeposit {
+  type: "Deposit";
+  policyId: PolicyId;
+  deposit: Deposit;
+
+  applyAction(): AppliedActionResult;
+}
+
+interface CanChoose {
+  type: "Choice";
+  policyId: PolicyId;
+  choice: Choice;
+
+  applyAction(choice: ChosenNum): AppliedActionResult;
+}
+
+/**
+ * This Applicable action is intended to be used when the contract is not in a quiescent state.
+ * This means that the contract is either timeouted, or it was just created and it doesn't starts with a `When`
+ */
+interface CanAdvance {
+  type: "Advance";
+  policyId: PolicyId;
+  applyAction(): AppliedActionResult;
+}
+
+export type ApplicableAction = CanNotify | CanDeposit | CanChoose | CanAdvance;
+
+function getApplicant(action: ApplicableAction): ActionApplicant {
+  switch (action.type) {
+    case "Notify":
+    case "Advance":
+      return "anybody";
+    case "Deposit":
+      return action.deposit.party;
+    case "Choice":
+      return action.choice.for_choice.choice_owner;
+  }
+}
+
+export async function getApplicableActions(
+  restClient: RestClient,
+  contractId: ContractId,
+  environment?: Environment
+): Promise<ApplicableAction[]> {
+  const contractDetails = await restClient.getContractById(contractId);
+  const currentContract = contractDetails.currentContract
+    ? contractDetails.currentContract
+    : contractDetails.initialContract;
+  const oneDayFrom = (time: Timeout) => time + 24n * 60n * 60n * 1000n; // in milliseconds
+  const now = datetoTimeout(new Date());
+  const nextTimeout = getNextTimeout(currentContract, now) ?? oneDayFrom(now);
+  const timeInterval = { from: now, to: nextTimeout - 1n };
+  const env = environment ?? { timeInterval };
+  if (typeof contractDetails.state === "undefined") {
+    // TODO: Check, I believe this happens when a contract is in a closed state, but it would be nice
+    //       if the API returned something more explicit.
+    return [];
+  }
+  const initialReduce = reduceContractUntilQuiescent(
+    env,
+    contractDetails.state,
+    currentContract
+  );
+  if (initialReduce == "TEAmbiguousTimeIntervalError")
+    throw new Error("AmbiguousTimeIntervalError");
+  const applicableActions: ApplicableAction[] = initialReduce.reduced
+    ? [
+        {
+          type: "Advance",
+          policyId: contractDetails.roleTokenMintingPolicyId,
```

Comment:
@hrajchert , I don't understand why `contractDetails.roleTokenMintingPolicyId` is not optional. can't we used a contract without role tokens ?